add sorting functionality for channels [wip]

spacelatte opened the pull request:
Hello,

I added channel-sorting functionality upon header-click.

I am not really obj-c/mac developer. So feedback is appreciated :)

Here is functionality in-action.

![example.gif](https://user-images.githubusercontent.com/5755154/74433994-8ef6d200-4e72-11ea-95a2-3b0e951e1044.gif)

Commit:
Significant changes to appearance

• TVCMainWindowAppearance is no longer the main appearance object. A lot of logic has been abstracted away into TXAppearance which is a singleton that keeps track of the application-wide appearance.
• The TVCApplicationAppearance subclass of TVCAppearance now exists which when used as a subclass of (such as for TVCMainWindowAppearance), autofills all properties from TXAppearance such as the appearance type, whether it is dark, etc.
• NSView category methods used for updating views when the appearance have moved to TXAppearancehelper.h.
• -applicationAppearanceChanged is now called instead of -mainWindowAppearanceChanged on each view when appearance changes.
• Added NSWindow category TXAppearanceHelper.h to make it easier for individual windows to inform its subviews of changes.
• Appearance change logic for TVCServerList and TVCMemberList have been greatly simplified. Much logic has been moved into a method name -invalidateBackgroundForSelection which is available in Cocoa Extensions framework.
• -isEmphasized now returns a different value depending on appearance.
• Other misc. changes not worth documenting

Commit:
Rename appearance preference

Commit:
Modify -isEmphasized logic

Commit:
Change binding I forgot to update

Commit:
No longer need reload theme & reload user interface combination

Table view appearance is no longer tied to the theme

Commit:
Modify channel spotlight appearance

• Remove default window height
• Added selection color for search result

Commit:
What is the point of weakSelf if notification is retaining self

Commit:
Rename method for consistency

Commit:
Move location of other copies check to main()

Commit:
Large refactoring of channel spotlight

• The channel spotlight is now its own window instead of being attached to the main window.
• The window now remains on screen unless escape is pressed or a channel is navigated to.
• When pressing escape, if there is a search string, that is cleared. The next press of escape then closes the window.
• The unread count and highlight count for each channel is now updated each time either change.
• Appearance now uses TVCAppearance and all drawing code has been modernized.
• The appearance of the view now correctly changes when channels and connections are deleted while they are in the current search results.
• Unpopular change: the window now closes when trial expires.

Commit:
oops

Commit:
Modify trial timers

When trial timers call, perform -stopTrialTimer to deallocate timer. It is one-shot timer.

Commit:
Add observer to channel spotlight for license deactivation

Commit:
Add TODO warning

Commit:
Use TXSystemIsOSXMojaveOrLater for Mojave only code

Commit:
Modify historic log manager

• We now hard code -includesPendingChanges and -includesPropertyValues to YES. For some reason, even though model was configured to set these to YES, they were set to NO instead. Need to build a demo app to see if it's a bug in Xcode or something specific to Textual.
• "GenericAll" fetch request template was removed because it's not used.
• "Truncate" fetch request template was modified to set -includesPropertyValues. We still hard code it for this fetch request, but having it in UI too at least gives me a reminder of my intetnions when I look at code and model.

Commit:
Reset value of -lastValidationErrorDescription when we have selection in combo box

Commit:
Change section title

Commit:
Modify TXAppearance

• All logic for changes has been condensed into a single method to simplfy a lot of logic that was unncessarily spaced out.
• -appKitLightAppearance and -appKitDarkAppearance are now declared as nullable.
• Replaced -appKitAppearanceInherited with -appKitAppearanceTarget which uses an enum to provide more granularity for how NSAppeaance objects are treated. On Mojave, when we are setup to inherit the system color, we now don't set an NSAppearance object on anything and let that be inherited from NSApp. -appKitAppearance is now nil when -appKitAppearanceTarget is none.

Commit:
Alter appearance of application icon

Commit:
Modify layout constraints

Commit:
Fix appearance changes by the system not registering correctly.

Commit:
Modify sample code

• Update Swift Bot sample to latest remote
• Remove DNSBL sample code because it's no longer maintained nor compatible with latest version of Textual.

Commit:
Modify TPCPathInfo

• Change -transcriptFolderURL to a property
• Add -transcriptFolder property
• Add missing "copy" attribute on all properties

Commit:
Big improvements to TLOFileLogger

• Log files are now closed after being idle too long (currently 20 minutes). So that we don't need a timer for each log file, we instead of a class wide timer under +[TLOFileLogger  idleTimer] that fires every 10 minutes. When this timer fires, it posts a notifications which instances can observe. A count of open file handles is kept track of to make sure we aren't doing extra work when no file handles are open. The idle timer runs in a global queue, with lowest possible priority.
• The path system has been reimagined to allow for greater consistency. -writePath is the folder in which log files for the item are written to. -filePath is -writePath combined with -fileName which is the file name of the log for the day. Internally, a -filePathHash property is now used to determine when important factors, such as the configured log location and -fileName, changes so that we know when to reopen the file handle.

Commit:
Add debug logging just cause

Commit:
Update frameworks to modules

Commit:
Modify project

• Add a little bit of swift just for the fun of it. The implementation files the swift replaces is gone but the header still remains because there isn't a good way to communicate swift interface as an application to a plugin. Apps don't have module maps.
• Renamed TextualApplication.h to Textual.h (private header too) just for the heck of it. No other reason than I thought it looked better.

Commit:
Modify loader path

Commit:
Update libraries

Commit:
Fix suppression info not being passed to parent

Commit:
Offer to enable dark mode if user loads a style that previously force it enabled

Commit:
Remove obsolete file reference

Commit:
Update projects to Xcode 10

Commit:
Fix close query missing keyboard shortcut

Commit:
Correctly save and restore preferences window position using helper

Commit:
Save and restore channel spotlight position on screen

Commit:
Fix incorrect condition

We shouldn't continue when there is a dark appearance

Commit:
-_-

Commit:
Transition core to use random IDs for localization in place of integer

If you're wondering why I don't follow the pattern of "<localization>" = "<localization>" which is more commonly accepted: it's because random keys make code cleaner. In my opinion. It also makes it easier to access different tables. TXTLS(), which is the function that returns strings, accepts first half of key as the table. For example "IRC[somekey]" translates to: lookup "somekey" in table "IRC" — There are others, but that is the most notable.

This commit also moves some strings to different files including the addition of a new file named CommonErrors.strings.

Commit:
Version bump

Commit:
Do not wrap label

Commit:
Modify logging formatters

Commit:
Update to new logging API in CE framework

Commit:
No need to call self

Commit:
Update frameworks

Commit:
Stop embedding swift in inline service

We get this for free from the app

Commit:
Rewrite connection manager in Swift

The entire connection manager service, except for the main entry point, has been rewritten in Swift. The connection interface is now modular by using subclasses. The reason it was written in Swift is because new Network.framework which there are plans to support doesn't have an Objective-C interface. There is only a plain C interface or Swift.

Commit:
Modify classic connection socket

• Keep track of disconnecting state when calling close()
• No longer declare a default implementation for readIn()
• Call willConnectTo delegate when connecting without a proxy

Commit:
Connection manager improvements

• The GCDAsyncSocketCipherNames and GCDAsyncTrustPanel class extensions have been extracted into their own classes (RCMSecureTransport and RCMTrustPanel) because the methods declared by these classes can be used universally, outside of GCDAsyncSocket.
• Many methods that aren't tried directly to GCDAsyncSocket have been moved out from the GCDAsyncSocketExtensions category into RCMSecureTransport.
• The readDelimiter property and Timeout enum have been moved into ConnectionSocketClassic because that's the only place they will be used.
• In ConnectionSocketClassic, the logic for responding to disconnects is moved to a separate function named onDisconnect(with:)
• In ConnectionSocketClassic, the logic for responding to secured connections is moved to a separate functioned named onSecured()
• In ConnectionSocketClassic, onSuccessfulConnect() is renamed to onConnect()
• ConnectionSocket now provides a default implementation for verifying certificates named tlsVerify(:response:)
• ConnectionSocketProtocol now specifies several TLS related properties
• ConnectionSocketProtocol now has a default implementation of exportSecureConnectionInformation() which use the aforementioned properties

Commit:
Add connectionPrefersModernSockets flag to config classes

Commit:
Add hidden preference to force modern sockets

Supersedes any flag set on IRCClient

Commit:
Use inline formatting with log messages

Commit:
Do not enforced `secured` property

This method may be called during validation in which case the connection hasn't been secured yet but we still have the relevant information

Commit:
Workaround for casting crash

Commit:
Add support for modern sockets aka Network.framework

Current limiations:
• Cannot set a custom proxy. Can only chose none or system.
• Cannot set to prefer IPv4

Disabled by default. While there is currently a per-client flag, it can only be enabled right now through hidden preferences.

Commit:
Weak link framework as it is new

Commit:
How'd that guy get there...

Commit:
Check disconnecting status in read/write handlers

Commit:
Respond to EOF

Commit:
Fix for Mavericks

Commit:
Modify -logFilePath logic

We no longer need to call -stringByDeletingLastPathComponent on the writePath for the log handler because the new implementation of TLOFileLogger already does this internally for us.

Commit:
TLOFileLogger write path can return nil value for utilities so we need to handle that

Commit:
Rename IRCConnection proxy type enum

Commit:
Replace "Prefer IPv4" preference

Because modern sockets do not support setting a preference for address type, we now make the address type explicit. User can pick to connect to only IPv4, only IPv6, or have it automatically chosen.

For users that had the old preference enabled, they will automatically be migrated to IPv4 only. Additionally, they are shown a warning each time they connect to inform them of the change in behavior. The warning disappears if they change the new preference, which is located in Network Socket section of Server Properties.

Commit:
Rename IRCConnectionProxyTypeSystemSocks

Commit:
Modify entitlements to correctly access preferences

Commit:
Automatically enable modern sockets

No reason to have a per-connection flag to enable or disable modern sockets.

If user is on an a version of macOS that supports it and they have none or automatic proxy type configured, then we use it.

Just in case there is some fallout for some unforeseen problem, a hidden preference will remain that can be used to turn it off app wide.

Commit:
Improve connection error handling

Commit:
Fix incorrect condition when writing to config

Commit:
Modify attributes for probably negligible optimizations

Commit:
Move location of extension

Commit:
Move error declaration

Commit:
Map domain and code of connection error using computed property

Commit:
Present error to user when XPC service exits unexpectedly

Commit:
Update frameworks

Commit:
Remove Swift Bot submodule in preperation for relocating it

Commit:
Reorganize project

Many resources have been added to Textual with little thought for long term maintenance. This commit is an overdue overhaul of the entire project. The goals of this overhaul was to introduce as much consistency as possible to the project.

Some notable changes:

• Code signing identity file has moved to "Configurations/Build"
• All build configuration files, including entitlements, moved to "Configurations"
• Source code for the app has moved to "Sources/App"
• Shared source code has moved to "Sources/Shared"
• Source code for extensions have moved to "Sources/Plugins"
• The project file for the app has moved along with its source code. It's recommended to use the Textual.xcworkspace workspace file.
• Build scripts are no longer packaged within the project files.
• Folders are now used throughout most projects instead of groups.
• Removed a lot of redundant framework linking.
• The preference panel sample plugin has been modified to use standard Cocoa facilities instead of those that are project specific. e.g. NSUserDefaults and NSLog() instead of TPCPreferencesUserDefaults and LogToConsole()

Commit:
Rename application icon set

Commit:
No need to include photoshop files in bulit product

Commit:
Reintroduce Swift Bot sample code at new location

Commit:
Only build on High Sierra or later

In theory user needs Xcode 10 at minimum already, which is limited to High Sierra or later. That's because we have Swift 4.2 specific code which will not compile on Xcode 9 or earlier.

For Xcode 9 support, see stable branch (master).

Commit:
Update README.md

Commit:
Modify update version build script

Don't keep making changes to files when we don't need to. Only make changes when something changes.

Commit:
Fix regression in /brag command because of localization changes

Commit:
Modify build configuration for all projects

Commit:
Stray semicolons

Commit:
Remove duplicate framework link

perryprog committed:
Add :thinking: to smiley converter

Commit:
Export certain headers as public

Commit:
Small code formatting change

Commit:
Update sample code

Commit:
Fixed arguments not passed to LogToConsole()

Commit:
Update frameworks

Commit:
Modify build configuration

Commit:
Fix (hopefully) building on macOS High Sierra

Commit:
Export headers as public

Commit:
Update frameworks

Commit:
Rename targets

This gives us a more consistent experience when auto generating schemes

Commit:
Modify preferences

• Removed unused defaults: "SyncPreferencesToTheCloudIncludesKeychainItems"
"Socket -> Secured Connection -> Enforce Allowed Cipher Suites"

• Renamed "GCDAsyncSocket Cipher List Includes Deprecated Ciphers" to "PreferModernCiphers" - This prefernece was hidden which means no migration logic will be added for the change.

• Removed "GenerateLocalizedTimestampTemplateToken" as it has outlived its stay (was used to diagnose a problem long ago).

Commit:
Forgot to change the build scripts in 86cc0b5

Commit:
Fix broken localization

Commit:
Use "en" as locale

Commit:
Rename targets

Commit:
Update Textual Extras installer

Package latest version of each extension

Commit:
Fix plurals

Commit:
Don't compile unused resources

Commit:
Modify plugin protocol

Removed implementation details for methods that have long since been considered deprecated. There is no reasonable explanation why a plugin may still implement these.

Commit:
Fix missing namespace in function call

Commit:
Remove dead code

Commit:
Refactor IRC text formatting

• IRCTextFormatterEffect is a brand new class which acts as a container for a single effect. Given an effect type and appropriate value, it is intelligent enough to properly parse the value into a string, assign a control character, and determine the number of bytes needed to append the effect to a string. It exposes two methods which can be used to append opening and closing to a mutable string.
• IRCTextFormatterEffects is a brand new class which acts as a container for all IRCTextFormatterEffect objects within a specific attributed string. It takes a dictionary of attributes as an input and spits out a new instance of self. Similar to IRCTextFormatterEffect, it exposes two methods which can be used to append opening and closing to a mutable string for all effects it contains.
• Almost the entire logic of the logic for truncating formatted text has been rewritten to be much more reliable and sensible. Constants have been changed to values that are much more appropriate for their purposes instead of just using numbers pulled out of thin air.
• Methods for formatting strings now takes a string channel name instead of IRCChannel so that even if the channel doesn't exist yet, we can still properly truncate.
• Methods for formatting strings have been renamed because I like the new names more. No other reason.

Commit:
Rename variables to match other implementation

Commit:
Make sure notification control menu items are always enabled

Commit:
Remove dead code

Commit:
Make static analzyer happier

• In cases where we know for sure that nullability warning is ignorable, cast to non-null.
• Added TXLocalizationNotNeeded() which inlines a string to represent it as localized even though it's not in reality.
• Nullability is now specified for some methods that can in fact return nil but were previously not annotated as such.

Commit:
Do not cast to void

These values are all discardable and we don't warnings for this.

Commit:
Do not generate debug symbols outside of Debug build

Commit:
Move IRCChannelUser to Users subfolder

Commit:
Refactor localization system

• Added TLOLocalization.swift which is a Swift implementation of the localization system. TXTLS() and its sister functions are bridged to this implementation.
• The TLOLanguagePreferences class no longer exists. This class was used only as a convenience to TXTLS() and its sister functions. I don't know of any and/or why a plugin may access these, so they were not deprecated.
• The TLOLanguagePreferences files have been renamed to TLOLocalization as the latter better suites its purpose.

Commit:
Fix incorrect error matching

Really should stop hard coding these...

Commit:
Modify -descriptionForErrorCode:

The method is now nonnull and returns "Unknown" for out of range error codes

Commit:
Improved connection error handling

• ConnectionError now inherits from CustomNSError and LocalizedError instead of rolling our own implementation.
• ConnectionSocketNWF now implements initializers for the various errors that the Network framework produces. This includes a big switch statement for DNS errors which don't have a native error string function.

Commit:
Fix building on macOS High Sierra

• firstIndex(of:) for Array has been moved to Cocoa Extensions
• firstIndex(of:) for String has been added to Cocoa Extensions to allow TLOLocalizations.swift to build on macOS High Sierra (Swift 4.1)

Commit:
Use correct attribute for Swift 4.1

Commit:
No longer hard code connection error information

Commit:
Fix incorrect copyright year caused by lazily copy and pasting

Commit:
Revert to calling -recalculateTextViewSizeForced in next main queue pass

I was doing too much work in -updateConstraints

Commit:
Modify spell checker for main text field

The main text field now exposes -resetSpellingIgnores in its private header which resets its ignored words list to self.defaultSpellingIgnores.

self.defaultSpellingIgnores returns contents from StaticStore.plist.

Current ignores:
"deop" so it stops auto correcting to "/cs drop"

Commit:
Fix for subreddit names -> link

Commit:
Update frameworks

Commit:
Update CE to inline some helpers

Commit:
newBuffer can never be nil thanks to line 215

Commit:
Use first(where:) instead of filter when we only need first result

Commit:
Access self for secured connection information

Commit:
Use weak reference to self in closures

Commit:
Do not use -hasSuffix: when matching domain names for inlining content.

Using -hasSuffix: makes logic easier, but it also introduces potential security holes in which an unrelated domain can masquerade as a legitimate service, which can lead to unintended consequences.

The ICLHelpers header of the inline content service now exposes two new NSString helpers:
• -isDomainOrSubdomain:   returns YES if self is equal to input or is a subdomain of input.
• -isDomain:              returns YES if self is equal to input.

Commit:
Plugins don't require entitlements

Plugins inherit whatever entitlements the host have because they are loaded into the host.

Commit:
Change bundled extensions and ICL plugins to first class citizens

We control the codebase of these things so why make them search for headers and frameworks in the compiled result of their host

Commit:
Add inline support for speedtest.net results

Commit:
Improve inline content error handling

ICLMediaAccessor now exposes a +logError: method which a caller can chose to call. The media accessor doesn't call it itself so that the module can decide to. The module to load all content chooses not to.

Commit:
Change pragma options for scrollback database

Commit:
Modify access control

Commit:
Update model list

Commit:
Use localizedString() in place of NSLocalizedString()

We access localized strings A LOT and this is one less hop

Commit:
Variable is not accessed outside of if statement so move it

Commit:
Add -lastLine property to view controller and proxy out through parents

-newestLineNumber will not be replaced by -lastLine because they are assigned under difference conditions and are used for different purposes.

-newestLineNumber is assigned when print operation is performed.
-lastLine is assigned when print operation is enqueued.

-lastLine gives us a great deal of information about the last line which we only need to know about until the next print operation is enqueued. After the print operation is performed, that information is no longer necessary. If it ever comes up again, we can access the database for the details, which we only need the line number to do.

Commit:
Add -isFirstForDay property to TVCLogLine

-isFirstForDay returnes YES if is first line for the day defined by receivedAt. IRCClient uses the -lastLine property added in fe781a2 to compare next line to last line to know when day changes. This reduces complexity as we aren't trying to figure out when date changes, live, which is just a whole lot of overhead.

Commit:
Remove code from first implementation of fe781a2 I overlooked

Commit:
Add class forward

Commit:
Assign lastLine (if there is none) when reloading history

Commit:
Modify session indicator

Renamed to session indicator as even for me, message buffer session indicator too verbose.

The session indicator is now added to the view using a template partial instead of other logic.

Commit:
Add encryption lock icon to view using template partial

Commit:
Create copy of results dictionary when assigning

Commit:
Fix incorrect condition for assigning isFirstForDay

Commit:
First generation indicator for when date changes ("date indicator")

Commit:
Remove date indicator logic from Equinox as its now built in

Commit:
Add preference to control whether date indicator is shown (on by default)

Commit:
Core JavaScript improvements

Almost the entire core.js file has been rewritten to provide greater documentation.

app.channelIsJoined() has been deprecated in favor of app.channelIsActive().
app.channelIsActive() will return a value for a regular channel or private message, so it makes more sense to use "active" over "joined"
Why don't I add app.privateMessageIsActive() instead? Because app.channelIsActive() can kill two birds with one stone.

app.channelMemberCount() now throws a JavaScript exception when used outside a channel.

Commit:
Expose -[IRCChannel addUser:]

Commit:
Add class forward for IRCUser

Commit:
Move to current best practices for enums and constant strings.

For enums, place case at end instead of in middle. For example: IRCClientConnectNormalMode turns into IRCClientConnectModeNormal

For constant strings, when appropriate, use an extensible string enum. Either one of those from Foundation (NSNotificationName, NSErrorDomain) or typedef our own.

Commit:
Always respond to CTCP LAGCHECK regardless of ignores or CTCP is disabled

We are only responding to ourselves so there is no reason not to allow this.

CTCP LAGCHECK is used by the "/lagcheck" and "/mylag" commands

Commit:
Fix notification mute not working

Regression introduced by c70d968744063f1cb2fa4f2f6f8b13556ddd60e8

Commit:
Alignment fix for unread badges on beta 4 of Mojave (retina only)

Commit:
Modify Info.plist files

• Some sections have been rearranged. Sections that collapse into a dictionary or array have been moved to the bottom, grouped together.
• Added entry for NSContactsUsageDescription

Commit:
Temporary? fix for laying out dynamic views on beta 4 of Mojave

Commit:
Always assign nil to self.appearance when appKitAppearanceTarget is not applicable

When a specific appearance is used, self.appearance may be assigned an NSAppearance object. When system appearance is used, the target specifies that self.appearance shouldn't contain a value so that it can be inherited, but because we were only setting self.appearance when appKitAppearanceTarget is applicable; the NSAppearance object already set was never unset.

Commit:
Add support for color digits 16 through 98

The colors are available through the color panel that can be accessed by clicking "Other…" in color menu

Commit:
Resolved by rdar://41652556

Commit:
Alignment correction for Mojave

Commit:
Add -[IRCClient resetSASLNegotation]

Commit:
Call -processPendingCapability: for CAP NEW

Previous behavior was to call -toggleCapability:enabled:isUpdateRequest:
While this accomplished the goal of enabling the capability, it provided no
acknowledgement to the server in the form of a REQ command.
-processPendingCapability: is the same method LS command is fed through
and through testing, there appears no harm in using it for NEW as well.

Commit:
Change order of condition

Commit:
Modify commands

Allow most commands to be used with a utility window selected.

Commit:
Add warning TODO

Commit:
Attempt to resolve "Close Query" keyboard shortcut issue

Only one menu item can have a keyboard shortcut which means because "Delete Channel" and "Close Query" are competiting for the same, there must be a way to determine which wins. Previous solution was to archive the keyboard shortcut set and restore it. This proved to be prone to failure. New solution is much simpler. We simply unassign the submenu when the parent menu is hidden so that the only time the keyboard shortcut is in the main menu is when it's the one we want to use.

Commit:
Add auto join delay timer preference

On a certain server, even though we received message from NickServ that identification was successful, joining channels failed because of lag between NickServ and the IRCd. This commit adds a new preference to the Flood Control section of Preferences which controls delay before auto join is performed. It defaults to 0.0 (no delay). It can be configured up to 10.0 seconds.

To accommodate this new feature, the current auto join timer has been renamed to auto join next join timer.

Commit:
Remove TODO

Commit:
Add TVCMainWindowSelectionChangedNotification and use it in a few places

Commit:
Update predicate for search results when conditions change

• Update predicate when main window selection changes and when preferences changes because it can be dependent on per-server configuration.

Commit:
Remove TODO

Commit:
Fix for "Channel" menu in WebKit view

 2a5296552228c739c5a6945f739be4e1d759706e changed property assignment and the changes were not thoroughly tested

Commit:
Move to Swift 4.2 only now that Xcode 10 is GM

Commit:
Remove unnecessary fix

Commit:
Use ARC with structs now that we only support Xcode 10 GM

Commit:
Change append syntax to make it look nicer

Hopefully original bug is fixed

Commit:
Add nil check before calling -fileURLWithPath:

Commit:
Revert "Change append syntax to make it look nicer"

This reverts commit ec903291467490eee9102d7d3cd21752b901da7f.

Bug is not resolved

Commit:
Update Sparkle

Commit:
Fix incorrect binding for preferences

Commit:
Modify UI - Auto layout correctoins

Commit:
Use Swift friendly TVCLogLineType enum name

Commit:
Repackage Textual Extras extensions

The user interface for the extensions were modified in 6086a3fd2823ee83f8dea3dcfacf2a1b8cf6ffd3

Nothing else changed.

Commit:
Version bump

Commit:
Modify auto join logic

05730ff8734c8afdc23ba3aa94e5db57b4e9d088 introduced bug because i did not test it thoroughly.

When the auto join delay timer fired, it called out to join the next channel, bu tit never started the timer to continue after doing that. I only tested the change on one channel whichis why I never caught this.

Commit:
Version bump

Commit:
Do not link App Store version against Sparkle

Commit:
Remove i386 from Growl framework

Commit:
Fix incorrect localization tokens

Commit:
Version bump

Commit:
Merge branch 'dev'

Commit:
Version bump

Commit:
Add warning TODO

Commit:
Favor for a friend

jwheare committed:
Simplify max nicklen checks for hostmask parsing

This removes checking isupport NICKLEN for `isHostmaskNicknameOn` (used for sender/prefix parsing) in all cases.

Reasons:
* This was explicitly disabled in several cases anyway (missing isupport entirely, znc)
* This would fallback to a default of 9 when no NICKLEN was provided in isupport, but a default of 50 when explicitly disabled
* NICKLEN is not strictly the maximum length of a nickname on an IRC server. See:
** https://tools.ietf.org/html/draft-hardy-irc-isupport-00#section-4.14 "A client elsewhere on the network MAY use a nick length of higher value"
** https://tools.ietf.org/html/draft-brocklesby-irc-isupport-03#section-3.13 "   This parameter does not restrict the length of any nicknames other clients on the network may use"
** https://defs.ircdocs.horse/defs/isupport.html#nicklen "Other clients on the network may have nicknames longer than this."

The most serious issue here is for servers *with* ISUPPORT but no NICKLEN key, with nicks longer than 9 chars, where parsing just silently breaks and stuff like server-initiated JOINs fail to register.

This now always uses a limit of 50, which is arbitrary but probably a safe enough sanity check.

Commit:
Add hackint to list of IRC networks

emsquared committed:
Merge pull request #523 from jwheare/patch-1

Simplify max nicklen checks for hostmask parsing

emsquared committed:
Merge pull request #524 from mweinelt/pr-hackint

Add hackint to list of IRC networks

remi6397 committed:
Load executable scripts with any extension

remi6397 committed:
Fix indentation

remi6397 committed:
Correct errors in script attributes check

williamleuschner committed:
Only confirm quitting if connections are active

williamleuschner committed:
Whoops, use tabs instead of spaces

xeenon committed:
Adopt supported-color-schemes CSS property for styles to get proper scrollbar appearance in 10.14.4 betas.

Commit:
Change condition making certificate trust panel never appear

-isConnected will never be true because evaluation will occur before the connection has been established for modern sockets.

Commit:
Update list of macintosh models

Commit:
Merge branch 'xeenon-supported-color-schemes' into dev

Commit:
Merge branch 'remi6397-script-executables' into dev

Commit:
Modify condition and log message for executable script files

Commit:
Version bump

williamleuschner committed:
Partially revert c6504130c8f8 to remove XIB changes

Commit:
Change App Store profile

Commit:
Merge branch 'FEATURE/conditional-confirm-quit' of https://github.com/williamleuschner/Textual into williamleuschner-FEATURE/conditional-confirm-quit

Commit:
Merge branch 'williamleuschner-FEATURE/conditional-confirm-quit' into dev

Commit:
Stop abusing private APIs for "Inspect Element" menu item

 When WebKit2 first announced, there was no way for an
 app to take an NSMenu, modify it, and return the result
 when a context menu is presented.

 WebKit1 had a delegate method for that named:
 -webView:contextMenuItemsForElement:defaultMenuItems:

 WebKit2 added this delegate method, in private, but it
 was not available until the next update of macOS.

 'Til then, we had to fake "Inspect Element" because we
 did not have access to the default implementation that
 is available through the delegate method.

 Apple now flags the function names that were used to do that.

 This will break "Inspect Element" on one version of macOS.
 Which one I don't even recall.
 But it is worth it to make the review process smoother.

eskimo committed:
Update IRC.strings

Lowercase the word "slow" in "Very Slow" to match the others like "Pretty good", "Not bad", etc. Not sure why "slow" is capitalized.

Commit:
Update libraries

Commit:
Update project to Xcode 11 / Swift 5

Commit:
Add "Base" region for localization

Commit:
Improve Swift compatibility

Commit:
Return after calling code on main thread

We would duplicate the work when not on a main thread because the check didn't exit.

Commit:
Update library to recognize "Catalina" in "/sysinfo"

Commit:
Update MacintoshModels.plist

Commit:
Version bump

emsquared committed:
Merge pull request #529 from eskimo/patch-4

Update IRC.strings

Commit:
Change condition making certificate trust panel never appear

isConnected may be true and isConnecting false in some cirumstances when asking for trust. There are too many what-ifs at this point so we just do it no matter what and if the connection returns no policy name, then there is nothing further we can do.

Commit:
Temporary? workaround for missing policy name

Commit:
Merge branch 'dev' of https://github.com/Codeux-Software/Textual into dev

Commit:
Add window in which custom CSS rules can be defined.

Because the rules added in this window are completely local to the client and has no influence from the outside, there is absolutely no sanitation performed. This allows for open abuse of the system by escaping from the <style> element to append JavaScript or anything else without making a local copy of the style.

Commit:
Prefer system font for custom commands text field

Commit:
Fix building Encryption Kit

Commit:
Reintroduce weights command for inspecting nickname completion

Commit:
Fix building Auto Hyperlinks

emsquared committed:
macOS Catalina compatible np script

emsquared committed:
Version bump

emsquared committed:
Declare document viewer ranks

spacelatte committed:
add sorting functionality for channels

spacelatte committed:
fix whitespace diffs

spacelatte committed:
allow re-ordering after sorting

pvtmert commented on Sources/App/Classes/Dialogs/TDCServerPropertiesSheet.m:
```diff
@@ -2305,6 +2305,15 @@ - (BOOL)tableView:(NSTableView *)tableView acceptDrop:(id <NSDraggingInfo>)info
 }
 
+- (void)tableView:(NSTableView *)tableView mouseDownInHeaderOfTableColumn:(NSTableColumn *)tableColumn {
+	NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:nil ascending:YES comparator:^NSComparisonResult(IRCChannelConfig *cfg1, IRCChannelConfig *cfg2) {
```
i dont know if i have to `dealloc` at the end of the function...

emsquared replied:
Textual uses [Automatic Reference Counting](https://en.wikipedia.org/wiki/Automatic_Reference_Counting) which is fancy speak for it not requiring `dealloc`. The compiler adds that for you.

emsquared commented:
Thank you for the pull request!

I will review the changes on my next day off.

pvtmert commented:
Thank you, i have discovered an issue yesterday.

Regarding `[self.channelListArrayController setSortDescriptors];`

If set, user cannot sort channels. (eg needs to save and re-open sheet)
If not set, sorting gets reverted.

I tried to re-assign `self.channelList` but its readonly property

spacelatte committed:
fix-ish sorting behaviour

1) header click sorts channels
  a) save -> saves sorted list
  b) dragging anything *restores old list*

2) in order to re-order sorted list;
  - user needs to click header, sort channels
  - save & exit
  - re-enter list to sort items
  - save again to persist stuff
